CenterForOpenScience/osf.io
# Moving State Machine to Sanctions [ENG-2128]

## Purpose


Moving the moderation workflow for Registrations to work on top of the existing Sanctions workflow and migrate all existing sanctions interactions to this workflow.
 
## Changes

Added a new set of states for sanctions objects (PENDING_ADMIN_APPROVAL, PENDING_MODERATOR_APPROVAL, ACCEPTED, ADMIN_REJECTED, MODERATOR_REJECTED, COMPLETE) and updated the sanctions class to operate as a state machine by inheriting from the new SanctionStateMachine.

On a given trigger ('approve'/'accept'/'reject'), the Sanction will validate that the caller of the trigger has authority to advance the state, update the new sanction_state field, and call a corresponding callback for the trigger (`_on_approve`, `_on_complete`, or `_on_reject`). These callbacks are largely unmodified from the previously existing functions of the same name for each sanction. Finally, the each trigger calls the new `update_moderation_state` function on its registration.

`update_moderation_state` updates the new `moderation_state` field on the registration, using the current sanction object and its state to determine the correct value. In cases where the state transition involves movement into or out of a state requiring moderator attention, `update_moderation_state` will also write a new RegistrationAction reflecting the transition.

Other than that, I updated lots of tests and scripts to reference this simplified flow. Lots of tests and scripts.

## QA Notes

Please make verification statements inspired by your code and what your code touches.
- Verify that existing registration_approval/embargo/retraction flows still work as normal

What are the areas of risk?
I updated the previously-out-of-date tests for our scripts to auto-approve sanctions after a certain amount of time, but those scripts breaking would seem to be not fun.

Any concerns/considerations/questions that development raised?

## Documentation



## Side Effects



## Ticket
